File: apps/backend-node/src/features/profile/profile.repository.ts
import { Injectable } from '@nestjs/common';
import type { profiles } from '@prisma/client';

import { PrismaService } from '../../prisma/prisma.service';

@Injectable()
export class ProfileRepository {
  constructor(private readonly prisma: PrismaService) {}

  async getById(profileId: string): Promise<profiles | null> {
    return this.prisma.profiles.findFirst({
      where: {
        id: profileId,
        deleted_at: null,
      },
    });
  }

  async update(
    profileId: string,
    data: {
      fullName: string;
      phone: string | null;
      preferredLanguage: string;
      marketingConsent: boolean;
      avatarUrl: string | null;
    },
  ): Promise<profiles | null> {
    const profile = await this.getById(profileId);
    if (!profile) {
      return null;
    }

    return this.prisma.profiles.update({
      where: { id: profileId },
      data: {
        full_name: data.fullName.trim(),
        phone: data.phone,
        preferred_language: data.preferredLanguage.trim(),
        marketing_consent: data.marketingConsent,
        avatar_url: data.avatarUrl,
        updated_at: new Date(),
      },
    });
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/profile/ProfileRepository.java
package com.housingplatform.profile;

import com.housingplatform.persistence.entity.Profile;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Repository;

@Repository
public class ProfileRepository {

  @PersistenceContext private EntityManager entityManager;

  public Optional<Profile> findById(UUID profileId) {
    return Optional.ofNullable(
        entityManager
            .createQuery(
                """
                select p from Profile p
                where p.id = :profileId and p.deletedAt is null
                """,
                Profile.class)
            .setParameter("profileId", profileId)
            .getResultStream()
            .findFirst()
            .orElse(null));
  }

  public Profile update(
      UUID profileId,
      String fullName,
      String phone,
      String preferredLanguage,
      boolean marketingConsent,
      String avatarUrl) {
    int updated =
        entityManager
            .createNativeQuery(
                """
                update public.profiles
                set full_name = :fullName,
                    phone = :phone,
                    preferred_language = :preferredLanguage,
                    marketing_consent = :marketingConsent,
                    avatar_url = :avatarUrl,
                    updated_at = timezone('utc', now())
                where id = :profileId and deleted_at is null
                """)
            .setParameter("fullName", fullName)
            .setParameter("phone", phone)
            .setParameter("preferredLanguage", preferredLanguage)
            .setParameter("marketingConsent", marketingConsent)
            .setParameter("avatarUrl", avatarUrl)
            .setParameter("profileId", profileId)
            .executeUpdate();
    if (updated == 0) {
      return null;
    }
    return findById(profileId).orElse(null);
  }
}
